### PR TITLE
Origin/idempotency core

### PR DESCRIPTION
**Pull request by KevenFuentes9**
### Related issues, RFCs

**Issue number:** 

### PR status

***Is this ready for review?:*** NO  
***Is it a breaking change?:*** NO

## Checklist

- [ ] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [ ] I have performed a *self-review* of my own code
- [ ] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [ ] I have made corresponding changes to the *documentation*
- [ ] I have made corresponding changes to the *examples*
- [ ] My changes generate *no new warnings*
- [ ] The *code coverage* hasn't decreased
- [ ] I have *added tests* that prove my change is effective and works
- [ ] New and existing *unit tests pass* locally and in Github Actions
- [ ] Any *dependent changes have been merged and published* in downstream module
- [ ] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/idempotency/src/IdempotencyHandler.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import { IdempotencyOptions } from './IdempotencyOptions';
+import { IdempotencyRecord } from 'persistence';
+import { IdempotencyInconsistentStateError, IdempotencyItemAlreadyExistsError, IdempotencyAlreadyInProgressError, IdempotencyPersistenceLayerError } from './Exceptions';
+
+export class IdempotencyHandler<U> {
+
+  public constructor(private functiontoMakeIdempotent: AnyFunctionWithRecord<U>, private functionPayloadToBeHashed: unknown, 
+    private idempotencyOptions: IdempotencyOptions, private fullFunctionPayload: Record<string, any>) {}
+
+  public determineResultFromIdempotencyRecord(idempotencyRecord: IdempotencyRecord): Promise<U> | U{
```
Comment (jeffrey-baker-vg) [inline packages/idempotency/src/IdempotencyHandler.ts:13]: This might be work for another PR, but we need to add comments to the public-facing methods that document how we expect the methods to be used

File: packages/idempotency/src/IdempotencyHandler.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import { IdempotencyRecord } from 'persistence';
+import { IdempotencyInconsistentStateError, IdempotencyItemAlreadyExistsError, IdempotencyAlreadyInProgressError, IdempotencyPersistenceLayerError } from './Exceptions';
+
+export class IdempotencyHandler<U> {
+
+  public constructor(private functiontoMakeIdempotent: AnyFunctionWithRecord<U>, private functionPayloadToBeHashed: unknown, 
+    private idempotencyOptions: IdempotencyOptions, private fullFunctionPayload: Record<string, any>) {}
+
+  public determineResultFromIdempotencyRecord(idempotencyRecord: IdempotencyRecord): Promise<U> | U{ 
+    if (idempotencyRecord.getStatus() === IdempotencyRecordStatus.EXPIRED) {
```
Comment (jeffrey-baker-vg) [inline packages/idempotency/src/IdempotencyHandler.ts:14]: I notice the Python version has comments explaining why this is an inconsistent state. It may be helpful to add similar comments here for context.

File: packages/idempotency/src/IdempotencyHandler.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      throw new IdempotencyInconsistentStateError();
+    } else if (idempotencyRecord.getStatus() === IdempotencyRecordStatus.INPROGRESS){
+      throw new IdempotencyAlreadyInProgressError(`There is already an execution in progress with idempotency key: ${idempotencyRecord.idempotencyKey}`);
+    } else {
+      // Currently recalling the method as this fulfills FR1. FR3 will address using the previously stored value https://github.com/awslabs/aws-lambda-powertools-typescript/issues/447
+      return this.functiontoMakeIdempotent(this.fullFunctionPayload);
```
Comment (jeffrey-baker-vg) [inline packages/idempotency/src/IdempotencyHandler.ts:20]: Does this achieve idempotency (FR1) or does it just prevent concurrent calls from executing the same code?

File: packages/idempotency/src/types/AnyFunction.ts [hunk 4]
```diff
@@ -1,6 +1,11 @@
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-type AnyFunction<U> = (...args: Array<any>) => Promise<U>;
+type AnyFunctionWithRecord<U> = (record: Record<string,any>) => Promise<U> | U;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type AnyIdempotentFunction<U> = (record: Record<string,any>) => Promise<U>;
 
 export {
-  AnyFunction
+  // AnyFunction,
```
Comment (vgphoenixcampos) [inline packages/idempotency/src/types/AnyFunction.ts:8]: Probably should be removed in the future.

File: packages/idempotency/src/IdempotencyHandler.ts (hunk 3 above)
Comment (jeffrey-baker-vg) [reply packages/idempotency/src/IdempotencyHandler.ts:20]: The main question I have is: should we recall the method or should we just not return a response. I don't recall whether we discussed this earlier.

**Review by jeffrey-baker-vg — APPROVED** [review]
Still curious about the answers to my question, but nothing here should stop it in my opinion
